Update texttest expectation and fixtures to match spec

jonaskoperdraat opened the pull request:
The spec clearly says `"Conjured" items degrade in Quality twice as fast as normal items`. However, the texttest stdout.org shows a conjured item decreasing in quality by 1 each day. 

Merely updating `stdout.gr` would omit the test case where the `sellIn` date is past due for a conjured item, so I've also updated all the test fixtures I could find to include that case, and updated `stdout.gr` accordingly.

emilybache commented:
I think you missed the point entirely. Approval tests are generated from actual program output. You almost never edit them by hand. These outputs reflect what the system actually does not what you want it to do in future.

jonaskoperdraat commented:
I don't think I did :). I think the approval tests are using output from an invalid program. 

I appreciate the fact that approval tests are generated from actual program output. As such, they can be used to validate an application after refactoring. However, the output of the actual program should still conform to the spec, right? If you record the output of a faulty program, and then approve that output as validation for a new implementation, then that new implementation can never be correct.

Also, being able to sell "conjured" items is something that has to be added, so that behavior is not captured from the original application, but from a new implementation. Which, as far as I can see, is incorrect.

If I did miss the point, then please, can you explain how an application that decreases the quality of a "Conjured Mana Cake" just as fast as a "+5 Dexterity Vest" (a regular item) conforms to the spec that says _"Conjured" items degrade in quality twice as fast as normal items_? If I understand the specs correctly, a "Conjured Mana Cake" that has 6 quality on day 0, should have 4 quality on day 1. Not 5 as listed in the original `stdout.gr`

I updated `output.gr` to reflect the workings of a correct program. And to ensure coverage of both the case that the `sellIn` date has passed or not. I also updated the initial situation in each fixture. All of course after I did the Kata, and confirmed the working of my own application with unit tests. I came across this inconsistency when verifying my output with the texttest output.

emilybache commented:
This repo is a starting position for an exercise. The starting code does not implement conjured items. It says so in the introduction, that is part of the exercise. The approval test that comes with the starting code reflects the program as it is implemented.

You do not do approval testing by first generating a correct output by hand then writing a program that produces it. That is the point you have missed.

jonaskoperdraat commented:
Ah, I see now. Thanks for clearing that up. So the approval test does not reflect the new situation, merely the old situation. In that case the `stdout.gr` makes perfect sense. I assumed the approval tests to be verification of application as it was intended to be. 

Would it make sense to add an approval test for that situation? That would allow devs to verify their implementation after doing the Kata. It would be like having validation of both the 'before' and 'after' state.

emilybache commented:
As I said, you do not create approved output for functionality that does not exist yet. What the application is intended to be will be negotiated between developers, testers and business people. I do not want to provide that in advance, I want people to think for themselves in this exercise.

jonaskoperdraat commented:
In that case, approval tests can only be used after the fact, and to ensure an applications output doesn't change under refactoring. That seems a bit limiting to me as I hardly ever _only_ refactor. 

Also, as this is an exercise, wouldn't it be nice if the dev performing the excercise can verify that they've done so correctly? That was the scenario I thought (incorrectly as it seems) I could use these tests for.

emilybache commented:
This is not a good forum for me to teach you about approval testing :-)

As an exercise, I want this to be more similar to real legacy code development. In real life you rarely get a detailed test case provided by the customer to tell you when you're done. So no, I don't want people to have a test like that in this exercise.

jonaskoperdraat commented:
Ok. You keep your secrets, then! 😉
